### PR TITLE
feat: validate generation values against site capacity

### LINKED ISSUE #191: Warning" If user uploads generation data > capacity of syste,
We would like to warn the user if they upload generation data which is larger than their site capacity. 

We could use a 102 http response 

https://http.cat/status/102

### PR DESCRIPTION
**Pull request by anxkhn**
# Pull Request

## Description

- Implemented logic to compare uploaded PV generation values against the site's capacity.
- Raised HTTP 102 error if any generation value exceeds the site's capacity, with a custom error message informing the user.
-  Added site capacity check in the /sites/{site_uuid}/pv_actual route.

Fixes #191 

## How Has This Been Tested?

- [x] Yes

_If your changes affect data processing, have you plotted any changes? i.e. have you done a quick sanity check?_

- [x] Yes

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by anxkhn** [comment]
Hey @peterdudfield !
Here's my PR. This is my first time contributing to github.com/openclimatefix/pv-site-api
Can you help set it up properly locally, with pv-site-datamodel so I can check API and do tests locally too?

Thanks!

File: tests/test_generation.py [hunk 1]
```diff
@@ -146,25 +146,43 @@ def test_post_pv_actual(db_session, client, sites):
...
+    # make iteration of pv values for one day at a specific site
+    pv_actual_iteration_above = MultiplePVActual(
+        site_uuid=str(site_uuid), pv_actual_values=[pv_actual_above_capacity]
+    )
+
+    # this makes sure the datetimes are iso strings
+    pv_actual_dict_above = json.loads(pv_actual_iteration_above.json())
+    response_above = client.post(f"/sites/{site_uuid}/pv_actual", json=pv_actual_dict_above)
+
+    assert response_above.status_code == 102, response_above.text
```
Comment (peterdudfield) [inline tests/test_generation.py:185]: would you be able to split this test into 2. One, the original test, the second checking the new behaviour?

File: pv_site_api/main.py [hunk 2]
```diff
@@ -277,6 +278,15 @@ def post_pv_actual(
         )
 
     generation_values_df = pd.DataFrame(generations)
+    site = get_site_by_uuid(session=session, site_uuid=site_uuid)
+    site_capacity_kw = site.capacity_kw
+    exceeded_capacity = generation_values_df[generation_values_df["power_kw"] > site_capacity_kw]
+    if not exceeded_capacity.empty:
```
Comment (peterdudfield) [inline pv_site_api/main.py:284]: I'd probably go for `len(exceeded_capacity) > 0` as I think its slightly easier to read than two negatives

File: pv_site_api/main.py [hunk 3]
```diff
@@ -277,6 +278,15 @@ def post_pv_actual(
...
+        raise HTTPException(
+            status_code=102,
+            detail=f"We processed your generation values, but noticed one (or more) values are "
+            f"larger than the site capacity of {site_capacity_kw} kW.",
+        )
```
Comment (peterdudfield) [inline pv_site_api/main.py:289]: this is great

**Review by peterdudfield — CHANGES_REQUESTED** [review]
Thanks so much for this @anxkhn, really appreviate you doing this. 

Ive suggested two fairly small changes, would you be able to do them?

Thanks again

**Comment by peterdudfield** [comment]
> Hey @peterdudfield ! Here's my PR. This is my first time contributing to github.com/openclimatefix/pv-site-api Can you help set it up properly locally, with pv-site-datamodel so I can check API and do tests locally too?
> 
> Thanks!

Hey. The PR looks good. 
1. You should be able to run tests locally with `poetry run pytest tests`, see readme
2. For running locally you can use `poetry run uvicorn pv_site_api.main:app --reload`. You can set `export FAKE=1` to make it run wiht fake data not real database. 
let me know how you get on

File: tests/test_generation.py (hunk 1 above)
Comment (anxkhn) [reply tests/test_generation.py:185]: I wanted to spilt the test into two, but I was unsure what would happen if I remove the original code function.

Comment (peterdudfield) [reply tests/test_generation.py:185]: what do you mean? 

You can just add another test (with a slightly differenet) name, and then it should run. 

I hope the current test we do, doesn't make generations above the capacicty, but if it does, we can adjust that too

**Comment by anxkhn** [comment]
Hey @peterdudfield 
I have updated as requested. 
Are there any more requirements?

**Comment by peterdudfield** [comment]
Sorry, we've been chatting on #191 . 

Would you be able to implement that. Its basically, 
1. If all power is below capacity, return 200 as normal
2. if any power value is above the capacity, but less than 2* the capacity, then return 200, but add warning in the Http response
3. if any power is above 2* the capacity, then return 400 with error message. 

Sorry if this issue changed a bit while you were developing it

**Comment by peterdudfield** [comment]
Yes status code 422 - thanks - https://fastapi.tiangolo.com/tutorial/handling-errors/#import-httpexception

For the values less than the capacity, the route can just be returned as normal. 

For values between 1 and 2 * the capacity, yes, the above method looks a good way to do it

**Comment by peterdudfield** [comment]
hi @anxkhn 

Just had some internal chats. 

1. Could we change the threshold to 1.1*capacity to cause the 422. 

2. Could this 1.1 be a environment variable, so we can change it really easily. The default should be 1.1. 
```error_generation_capacity_factor = float(os.getenv('ERROR_GENERTAION_CAPACITY_FACTOR', 1.1))```

3. 
The error should say something like 
```
Error processing generation values. One (or more) values are larger than {error_generation_capacity_factor} times the site capacity of {capacity} kWp. Please adjust this generation value, the site capacity, or contact quartz.support@openclimatefix.org."
```

4. Below 1.1, don't worry about the warning message. We now think people wont look at this. 

Sorry its all changing, really appreciate your help on this

**Comment by anxkhn** [comment]
@peterdudfield I have ignored the warning message for now.
Rest code and test is updated and tested to handle 1.1 capacity_factor.

Please review the PR

File: pv_site_api/main.py [hunk 4]
```diff
@@ -276,7 +277,26 @@ def post_pv_actual(
...
+
     generation_values_df = pd.DataFrame(generations)
+    site = get_site_by_uuid(session=session, site_uuid=site_uuid)
+    site_capacity_kw = site.capacity_kw
+    exceeded_capacity = generation_values_df[
+        generation_values_df["power_kw"] > site_capacity_kw * capacity_factor
+    ]
+    if len(exceeded_capacity) > 0:
+        raise HTTPException(
+            status_code=102,
```
Comment (peterdudfield) [inline pv_site_api/main.py:291]: could this be 422

File: tests/test_generation.py [hunk 5]
```diff
@@ -146,25 +146,52 @@ def test_post_pv_actual(db_session, client, sites):
...
-    assert str(generations[0].site_uuid) == str(pv_actual_iteration.site_uuid)
+    assert str(generations[0].site_uuid) == str(pv_actual_iteration_below.site_uuid)
+
+
+def test_post_pv_actual_above_capacity(db_session, client, sites):
+    db_session.query(GenerationSQL).delete()
+
+    site_uuid = sites[0].site_uuid
+    site_capacity_kw = sites[0].capacity_kw
+    capacity_factor = float(os.getenv("ERROR_GENERATION_CAPACITY_FACTOR", 1.1))
```
Comment (peterdudfield) [inline tests/test_generation.py:177]: It would keep it simple in the test, and just set this to 1.1

**Comment by peterdudfield** [comment]
> @peterdudfield I have ignored the warning message for now. Rest code and test is updated and tested to handle 1.1 capacity_factor.
> 
> Please review the PR

Looks great, and thank you so much for doing this so quickly. Two small comments and I think itll then be ready to go. 
Thank you

**Comment by anxkhn** [comment]
@peterdudfield Hi.
Review the changes and let me know if any more is needed. Thanks!

File: .python-version [hunk 6]
```diff
@@ -0,0 +1 @@
+3.11.10
```
Comment (peterdudfield) [inline .python-version:1]: do you mind removing this?

Comment (peterdudfield) [reply .python-version:1]: If we need this, we can put this in another PR im sure

File: pv_site_api/main.py [hunk 7]
```diff
@@ -276,7 +277,26 @@ def post_pv_actual(
...
+    site = get_site_by_uuid(session=session, site_uuid=site_uuid)
+    site_capacity_kw = site.capacity_kw
+    exceeded_capacity = generation_values_df[
+        generation_values_df["power_kw"] > site_capacity_kw * capacity_factor
+    ]
+    if len(exceeded_capacity) > 0:
+        raise HTTPException(
+            status_code=422,
+            detail=(
+                "Error processing generation values. "
```
Comment (peterdudfield) [inline pv_site_api/main.py:293]: Oh and I think we need a f-string here, so itll be `f"`. This makes the variables get resolves. Also on another couple of strings in this block

**Comment by anxkhn** [comment]
Hi @peterdudfield 
I have made the requested changes.

**Comment by peterdudfield** [comment]
Just one comment about .python-version file, could you remove that?

**Comment by anxkhn** [comment]
@peterdudfield Done.

**Review by peterdudfield — APPROVED** [review]

**Comment by peterdudfield** [comment]
Thanks so much @anxkhn

**Comment by peterdudfield** [comment]
@all-contributors please add @anxkhn for code

**Pull request merged**